### PR TITLE
[CINN] Add new group scheduler

### PR DESCRIPTION
**Pull request by BiynXu**
### PR types
New features

### PR changes
Others

### Description
card-74457

Add GroupScheduler to schedule fusion groups.
Its responsibility is to perform loop alignment,  automatic inline, automatic loop fusion, and optimize the storage location of intermediate variables.

### CODE REVIEW

**Review by zhhsplendid — DISMISSED** [review]
LGTM

File: paddle/cinn/auto_schedule/search_space/auto_gen_rule/reduction_factoring.cc [hunk 1]
```diff
@@ -161,22 +161,42 @@ void ReductionFactoring::Apply(const std::string& block_name,
   // 5. Split the reduction loop into 2 part
   VLOG(6) << "before Split: " << ir_schedule->GetModule().GetExprs()[0];
   int factor = 1;
+  int max_factor = 1024;
   int extent = ir::GetLoopExtent(fused_reduce_loop);
-  for (int i = ceil(sqrt(extent)); i >= 1; --i) {
+  for (int i = max_factor; i >= 1; --i) {
```
Comment (zhhsplendid) [inline paddle/cinn/auto_schedule/search_space/auto_gen_rule/reduction_factoring.cc:166]: Here we change the speed up logic from `ceil(sqrt(extent))` to `max_factor`, why?

Should it be `std::min(max_factor, ceil(sqrt(extent)))` ?

File: paddle/cinn/hlir/framework/group_scheduler.cc [hunk 2]
```diff
@@ -0,0 +1,1224 @@
...
+        load_indice_value, CudaBindInfo::kCudaThread, load_block_name);
+    Range store_serial_overall_range = CalculateRange(
+        store_indice_value, CudaBindInfo::kSerial, store_block_name);
+    Range load_serial_overall_range = CalculateRange(
+        load_indice_value, CudaBindInfo::kSerial, load_block_name);
+    auto store_thread_coefficient_and_range = GetCoefficientAndRange(
+        store_indice_value, ir::ForType::GPUThread, store_block_name);
+    auto load_thread_coefficient_and_range = GetCoefficientAndRange(
+        load_indice_value, ir::ForType::GPUThread, load_block_name);
+    VLOG(6) << "store_indice_value: " << store_indice_value;
```
Comment (zhhsplendid) [inline paddle/cinn/hlir/framework/group_scheduler.cc]: Can you just have some operator `<<` on `range`, so you don't have to handle log stream of range.min, range.max

File: paddle/cinn/hlir/framework/group_scheduler.cc [hunk 3]
```diff
@@ -0,0 +1,1224 @@
...
+            << load_thread_coefficient_and_range[1].second.max << ")>";
+    VLOG(6) << "store_thread_coefficient_and_range[2] = <"
+            << store_thread_coefficient_and_range[2].first << ", ("
+            << store_thread_coefficient_and_range[2].second.min << ", "
+            << store_thread_coefficient_and_range[2].second.max << ")>";
+    VLOG(6) << "load_thread_coefficient_and_range[2] = <"
+            << load_thread_coefficient_and_range[2].first << ", ("
+            << load_thread_coefficient_and_range[2].second.min << ", "
+            << load_thread_coefficient_and_range[2].second.max << ")>";
+    return !(store_thread_overall_range.min <= load_thread_overall_range.min &&
```
Comment (zhhsplendid) [inline paddle/cinn/hlir/framework/group_scheduler.cc:901]: These code has bad readability, do we have a way to simply it?

File: paddle/cinn/hlir/framework/group_scheduler.cc [hunk 4]
```diff
@@ -0,0 +1,1224 @@
...
+            << load_block_coefficient_and_range[1].second.max << ")>";
+    VLOG(6) << "store_block_coefficient_and_range[2] = <"
+            << store_block_coefficient_and_range[2].first << ", ("
+            << store_block_coefficient_and_range[2].second.min << ", "
+            << store_block_coefficient_and_range[2].second.max << ")>";
+    VLOG(6) << "load_block_coefficient_and_range[2] = <"
+            << load_block_coefficient_and_range[2].first << ", ("
+            << load_block_coefficient_and_range[2].second.min << ", "
+            << load_block_coefficient_and_range[2].second.max << ")>";
+    return !(store_block_overall_range.min <= load_block_overall_range.min &&
```
Comment (zhhsplendid) [inline paddle/cinn/hlir/framework/group_scheduler.cc:975]: Same as above

File: paddle/cinn/auto_schedule/search_space/auto_gen_rule/reduction_factoring.cc (hunk 1 above)
Comment (BiynXu) [reply paddle/cinn/auto_schedule/search_space/auto_gen_rule/reduction_factoring.cc:166]: Here, the reduce loop will be divided into two parts: (extent / factor) and (factor), with the first part performing a serial reduction and the second part performing a parallel tree reduction. The current simple strategy is to hope that the second part is as large as possible, which can improve parallelism and reduce the probability of accuracy errors caused by adding large and small numbers.
Of course, currently these strategies are relatively naive, and in the future, we need more complex strategies to improve performance.

File: paddle/cinn/hlir/framework/group_scheduler.cc (hunk 2 above)
Comment (BiynXu) [reply paddle/cinn/hlir/framework/group_scheduler.cc]: done

File: paddle/cinn/hlir/framework/group_scheduler.cc (hunk 3 above)
Comment (BiynXu) [reply paddle/cinn/hlir/framework/group_scheduler.cc:901]: After implementing auxiliary data structures such as IntegerSet and MultiDimIntegerSet, we can re implement this function to simplify these ugly codes. I have added a TODO above the implementation of function AllocateStorage().

File: paddle/cinn/hlir/framework/group_scheduler.cc (hunk 4 above)
Comment (BiynXu) [reply paddle/cinn/hlir/framework/group_scheduler.cc:975]: After implementing auxiliary data structures such as IntegerSet and MultiDimIntegerSet, we can re implement this function to simplify these ugly codes. I have added a TODO above the implementation of function AllocateStorage().

File: paddle/cinn/auto_schedule/search_space/auto_gen_rule/reduction_factoring.cc (hunk 1 above)
Comment (zhhsplendid) [reply paddle/cinn/auto_schedule/search_space/auto_gen_rule/reduction_factoring.cc:166]: Good

**Review by zhhsplendid — APPROVED** [review]
LGTM

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Pull request merged**